pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/MultiRolesTokenAuthorizationProvider.java

```diff
@@ -0,0 +1,231 @@
+
+    // When symmetric key is configured
+    static final String CONF_TOKEN_SETTING_PREFIX = "";
+
+    // The token's claim that corresponds to the "role" string
+    static final String CONF_TOKEN_AUTH_CLAIM = "tokenAuthClaim";
+
+    private JwtParser parser;
+    private String roleClaim;
+
+    public MultiRolesTokenAuthorizationProvider() {
+        this.roleClaim = Claims.SUBJECT;
+        this.parser = Jwts.parserBuilder().build();
+    }
+
+    @Override
+    public void initialize(ServiceConfiguration conf, ConfigurationCacheService configCache) throws IOException {
+        String prefix = (String) conf.getProperty(CONF_TOKEN_SETTING_PREFIX);
+        if (null == prefix) {
+            prefix = "";
+        }
+        String confTokenAuthClaimSettingName = prefix + CONF_TOKEN_AUTH_CLAIM;
+        if (conf.getProperty(confTokenAuthClaimSettingName) != null
+                && StringUtils.isNotBlank((String) conf.getProperty(confTokenAuthClaimSettingName))) {
+            this.roleClaim = (String) conf.getProperty(confTokenAuthClaimSettingName);
+        }
+
+        super.initialize(conf, configCache);
+    }
+
+    private List<String> getRoles(AuthenticationDataSource authData) {
+        String token = null;
+
+        if (authData.hasDataFromCommand()) {
+            // Authenticate Pulsar binary connection
+            token = authData.getCommandData();
+            if (StringUtils.isBlank(token)) {
+                return Collections.emptyList();
+            }
+        } else if (authData.hasDataFromHttp()) {
+            // The format here should be compliant to RFC-6750
+            // (https://tools.ietf.org/html/rfc6750#section-2.1). Eg: Authorization: Bearer xxxxxxxxxxxxx
+            String httpHeaderValue = authData.getHttpHeader(HTTP_HEADER_NAME);
+            if (httpHeaderValue == null || !httpHeaderValue.startsWith(HTTP_HEADER_VALUE_PREFIX)) {
+                return Collections.emptyList();
+            }
+
+            // Remove prefix
+            token = httpHeaderValue.substring(HTTP_HEADER_VALUE_PREFIX.length());
+        }
+
+        if (token == null)
+            return Collections.emptyList();
+
+        String[] splitToken = token.split("\\.");
+        String unsignedToken = splitToken[0] + "." + splitToken[1] + ".";
+
+        Jwt<?, Claims> jwt = parser.parseClaimsJwt(unsignedToken);
+        try {
+            Collections.singletonList(jwt.getBody().get(roleClaim, String.class));
+        } catch (RequiredTypeException requiredTypeException) {
+            List list = jwt.getBody().get(roleClaim, List.class);
+            if (list != null) {
+                return list;
+            }
+        }
+
+        return Collections.emptyList();
+    }
+
+    public CompletableFuture<Boolean> authorize(AuthenticationDataSource authenticationData, Function<String, CompletableFuture<Boolean>> authorizeFunc) {
+        List<String> roles = getRoles(authenticationData);
```

Comment:
`getRoles()` might introduce the `RequiredTypeException` if the value not the String type and List type. We should make sure the CompletableFuture can be handled correctly.